### ISSUE TITLE
Slow file events?

### ISSUE THREAD

**Issue by coderars** [glenn2223/vscode-live-sass-compiler#22]
Hi @glenn2223,
thank you for your work on keeping this plugin alive!

I've just switched to your fork in hope of some speed improvements but I still experiencing very slow file events. I mean every time I press save on an SCSS file there's ~4sec delay when the "Change detected..." message shows up in the output window. It was the same with ritwick's version.
 
Do you have any idea why this long delay exists?

I'm on Windows 10 and VS Code is running in WSL mode (WSL2/Ubuntu 18.04.5 LTS).
"Extension is enabled on 'WSL: Ubuntu'"

Usually, for bigger projects I use my own Webpack configuration - to compile css, js and all the usual stuff - and file watching is handled by [chokidar](https://www.npmjs.com/package/chokidar). With that, file events are almost immediate, so I think this long delay is not caused by OS or WSL settings.

**Comment by sKopheK** [comment]
hi @coderars, thanks for bringing it up once again, i've noticed the same behavior, but in my case it's probably due to git in VS Code, that also scans files for changes and just after that, sass live compiler can step in, which causes the significant delay
mentioned it [here](https://github.com/glenn2223/vscode-live-sass-compiler/issues/17#issuecomment-763705511) for a large project, but it happens in a project with a single (+ a couple of partials) scss file

**Comment by glenn2223** [comment]
Sorry to hear about your issue @coderars and thanks for sharing the insight @sKopheK.

I think I might have identified an issue where I check if the file that is saved is excluded by your settings. However, can you please complete the below steps, just to confirm/help me debug the issue?

Get debug info:
1. Press `F1` or `Ctrl + Shift + P`
2. type: `Developer: Show running extensions` an hit `Enter`
3. This will show a `Running Extensions` tab
3. Click the circle in the top right corner (see the third icon in figure 1)
4. Go to a file that was the worst (performance-wise) and save it (triggering the extension to compile)
5. Wait for it to compile and for the `Watching...` message to be displayed in the output
5. Return to the `Running extensions` tab and click the stop icon (which will have replaced the third icon from figure 1)
6. Click the save icon (see the first icon in figure 1)
7. A dialogue will open it, save it as `CPU-Old.cpuprofile`

<caption>Figure 1</caption>

![image](https://user-images.githubusercontent.com/33450392/105996316-10566d00-60a2-11eb-8985-add8fc111465.png)

Install this preview: [live-sass-4.3.5-alpha.1.zip](https://github.com/glenn2223/vscode-live-sass-compiler/files/5880511/live-sass-4.3.5-alpha.1.zip)
1. Unpack the zip folder
2. Drag the `live-sass-4.3.5-alpha.1.vsix` to the extension area of VS Code (see figure 2)
3. Run the `Get debug info` steps above this time saving the profile as `CPU-New.cpuprofile`

<caption>Figure 2</caption>

![image](https://user-images.githubusercontent.com/33450392/105996792-b7d39f80-60a2-11eb-9a5a-4b9a93ff1341.png)

After you've done this can you come back and tell me if there was a performance improvement and include the two `cpuprofile` files ( you may have to zip them up to attach to github)

**Comment by coderars** [comment]
Hmm... I've just made a very basic local test, and now it's super fast, watcher starts immediately:

![basic-test](https://user-images.githubusercontent.com/37109162/105988329-c4a2c400-609f-11eb-8ddb-0a2e8a5e9b5b.jpg)

**My real project is a WordPress project**, and the project root for VS Code is the Wordpress root directory with all its **9985 files**. So after my above testing I figured out that my reported problem is because the plugin starts the watcher for the whole VS project root directory and that's why it gets very slow with so many files to watch...

The `liveSassCompile.settings.includeItems` option does not limit the watcher to only watch the given directories, it's just for checking if the currently edited file should trigger the css compilation or not (fix me if not).

**The solution:**
Adding some big directories to `liveSassCompile.settings.excludeList` did solve my problem, now the watcher is real fast!

However I cannot write a working negative blob pattern, to exclude everything except files in my theme dir: `themes/mytheme` but I'm still trying (my fault I think, need to read more docs on negative patterns).

**Comment by coderars** [comment]
@glenn2223 thanks, I'll do your tests soon. Do you want my tests to run without the `excludeList` option, as it was when I reported the issue?

**Comment by glenn2223** [comment]
>The `liveSassCompile.settings.includeItems` option does not limit the watcher to only watch the given directories, it's just for checking if the currently edited file should trigger the css compilation or not (fix me if not).

This is right. Partials may be in another directory so I don't want to restrict watching to those specific directories/files
.

>Adding some big directories to `liveSassCompile.settings.excludeList` did solve my problem, now the watcher is real fast!

Glad you found a solution! Definitely works great in your case, but a genuinely large project with loads of folders each with sass files will still suffer.
.

>Do you want my tests to run without the `excludeList` option, as it was when I reported the issue?

Please, I want to see the improvement from the worst-case scenario. 

I also modified the script to still do the exclusion check but still be quicker (hopefully). Could you then install this file and get the debug info for this version as well? [live-sass-4.3.5-alpha.2.zip](https://github.com/glenn2223/vscode-live-sass-compiler/files/5880920/live-sass-4.3.5-alpha.2.zip) Sorry to ask, it's just best to get some real-world data sometimes 😄

**Comment by coderars** [comment]
@glenn2223 here are my test results: [vscode_cpuprofiles.zip](https://github.com/glenn2223/vscode-live-sass-compiler/files/5881409/vscode_cpuprofiles.zip)


With `alpha1` the plugin not fired after save (nothing showed up in the output)

With `alpha2` output shows "Change detected..." immediately, but  the next "Compiling" output arrives ~3-4sec after and it's still slow while the whole process finishes.

With `alpha2` or `old (4.3.4)` but with dirs added to exclude list are both version fast enough, I don't feel any difference.
````json
{
  "liveSassCompile.settings.excludeList": [
    "**/vendor/**",
    "**/web/wp/**",
    "**/app/languages/**",
    "**/app/mu-plugins/**",
    "**/app/plugins/**",
    "**/app/uploads/**",
    "**/themes/astra/**" 
  ],
}
````
My 2 scss files are very simple, less than 100 lines and they are not importing any other files. I still think the problem is with too many files sent to the watcher when the plugin starts. Look at this related Webpack option:
> For some systems, watching many files can result in a lot of CPU or memory usage.

https://webpack.js.org/configuration/watch/#watchoptionsignored

I can see the difference when I open up the Windows Task Manager and looking at the "Vmmem" process which is used by WSL2 (where actually the plugin is running in my case). When the plugin works after saving a file, CPU usage for this process is 35% without excludes versus only 5% using the above exclude list.

I think the best would be a new option, something like "liveSassCompile.settings.watchRoot" or only watching items that are in the "liveSassCompile.settings.includeItems" array.

**Comment by glenn2223** [comment]
Thanks for doing all that @coderars . A quick deep dive into the functionality: there is no actual watcher as such (only the inbuilt VS Code one), when a change is detected VS Code triggers some code in the extension. When the file is a partial SASS (in the fact it starts with a `_`) then we must find all files (using the `glob` package). This package is causing a bottleneck in larger code bases.

⚠️  **For now:** anyone with a large code base but few SASS files; use the `excludeList` to target or remove folders that should never need to be checked. This can be done folder by folder (unless you can capture a few with a glob pattern)

-----

**Large code bases with many SASS files**

I can see that the inbuilt `glob` package is causing performance issues in larger projects. I've been doing some reading and think I will move over to `fdir`. It still supports all the usual glob patterns with a one exception but added support for a couple of others. 

I will do some testing to make sure the change is worth while and then document all the differences ready for a new major release

**Comment by coderars** [comment]
Thank you @glenn2223 for the deeper explanation. I usually work with Webpack, but now I see "watching" means different here. Today I had some time to examine `appModel.ts` and now I understand what's going under the hood. Yes, `fdir` seems [ways faster](https://thecodrr.github.io/fdir/globbers.html), but IMHO the main reason for slowness is that the code could be better optimized for performance but I'm not a nodejs+typscript expert.

For a quick&easy solution and to keep backward compatibility I still hold my opinion to introduce a new option to set a base directory where scanning for files should start from. If that option is set, the default `vscode.workspace.workspaceFolders` could be reduced to a much smaller array and speed wouldn't be an issue anymore. Actually, only using this option would make `includeItems` and `excludeList` options unnecessary (in most cases). Also `compileOnSave()` could check if `currentFile` is under that base path and only start to work when `true`.

Basicly, doing this would require only one new function e.g. `getWatchFolders()` and then replace all `vscode.workspace.workspaceFolders.map` to `this.getWatchFolders().map`.

I think in 99% of cases we store our `.scss` files (and its partials) in e.g. `/xxx/yyy/src/sass` then it's enough to watch/scan this folder and its subfolders. Other `.scss` files outside this folder are only vendor packages we `@import` but those files shouldn't be edited (= not need to watch). However, in the remaining 1% this "base dir" option should be left empty or just `/xxx` (whatever the user wants)

**Comment by glenn2223** [comment]
>Yes, fdir seems ways faster

Yeah, all the online benchmarking looks amazing (but I don't like to commit to anything in a thread until I see that actual results for myself 😄)

>the default `vscode.workspace.workspaceFolders` could be reduced to a much smaller array...

The `vscode.workspace.workspaceFolders` field isn't actually the subfolders you currently have open. It's the root folders you have in your workspace.  
***Explanation:** click on `File` then you should see `Add Folder To Workspace...`. This feature basically means that you can have folders at `c://a/b/c` and `d://x/y/z` in the same workspace. So I have to search all workspace folders (wherever they are) for the matching glob files.*

>a new option to set a base directory

I do like this idea and have thought about it myself (`baseDirectoryFolder`). The only niggle I have is with multi-folder workspaces (as I describe above) there's no guarantee that you will have the exact same folder structure across the workspace folders.

Maybe we only allow it on single folder workspaces?? But then there's going to be issues about activating it across multi-folder workspaces, then a new option for `allowBaseDirectoryFolderOn`: `multi-folder` | `single-folder` comes in. All this just becomes more and more complicated when a faster directory crawler can be a simpler solution, *in terms of settings*.


***:warning: Open for discussion on all of this btw***

----

### Thoughts so far
- v5 release
- move to fdir
- possibly add new settings for an extension to the base directory
  - e.g. `src/css` on `c://a/b/` will shift glob/fdir's focus to the directory `c://a/b/src/css` (*thus negating all other folders*)

**Comment by coderars** [comment]
I do not really understand the problem you described (my fault) but I absolutely have no experience with VS Code plugin API. Actually, I've never ever used this "Add Folder to Workspace..." thing 😃 

However, I wouldn't afraid to keep that new `baseDirectoryFolder` option simple. It should be only a notice in the documentation for the multi workspace scenario. IMHO: it's again a 1% problem. And it's just an option. The user is responsible for what option with what values he set. If that option doesn't fit his workspace settings then he can just avoid that option and everything goes back to default.

As a new (more simple) idea, I not really understand why would it hurt anyone? (sketch only):
````js
vscode.workspace.workspaceFolders.map((folder) => {

  // if we are not iterested in this folder (compared to our new option)
  // then just continue (not running the glob)

  return glob(
    // the slow part now, when glob has to check thousands of folders
  )
})
````

If you want to go for sure, call the option something like `forceBaseDirectory` 😄  Where "force" means: you (the user) are responsible if something goes wrong. This is common practice in naming options.

So personally I would take the risk (which equals 0 for me) and would only use this one option, instead `includeItems` and `excludeList` with having lots of different dirs for different projects.

**Comment by glenn2223** [comment]
Thanks for all the help on this btw!!

>So the .map() runs only once if we have single-folder workspace, right?

That's right, only once

>If you want to go for sure, call the option something like `forceBaseDirectory`

Good shout, the force bit definitely make sense.

I'll add this to v4.4 tonight as an added feature and keep the `fdir` migration pushed to v5.

I'm going to do a quick check tonight but I'm sure that settings can be applied at both folder level and workspace level. This means that my worry of a multi-folder workspace having different structures would become null and void

>`//change glob's cwd option according to our new option (if set, instead folder.uri.fsPath)
  )`

I was thinking of appending rather than replacing. So it could be set to `"src/sass/"` from your example

*(we'll I say append but it'll build the path relative to `folder.uri.fsPath`)*

**Comment by glenn2223** [comment]
Moving the breaking change to a new issue

**Comment by glenn2223** [comment]
>I'm going to do a quick check tonight but I'm sure that settings can be applied at both folder level and workspace level. This means that my worry of a multi-folder workspace having different structures would become null and void

So I was wrong. And, after opening an issue with VS Code, I can see why (how will VS Code know which settings to use without a path - complicated setup if it went ahead)

**But I had a brain wave!** (a bit of an obvious one really 😄 )  
The setting currently allows either a `string` or `null`, I'm going to allow a string array too

It'll then search the array returning the first match.

I was also supposed to add a check of the file path to make sure it exists, but this must have slipped my mind before I released. I'll incorporate this too

**Comment by coderars** [comment]
Superb! Thank you! Now my config is simple and clean and the plugin works very fast.
There's only one little thing I think is a bit confusing:

For this new `forceBaseDirectory` option the documentation says that its value **MUST** start without a leading slash:

> No leading slash but MUST have ending slash
> Example: src/style/

However for the option `savePath` in `formats` it is:

> // / -> denotes relative to the workspace root
>         "savePath": "/dist/css"
> 

So using both of them looks confusing in `settings.json`:
````json
  "liveSassCompile.settings.forceBaseDirectory": "web/app/themes/child/src/sass/",
  "liveSassCompile.settings.formats": [
    {
      "format": "expanded",
      "extensionName": ".css",
      "savePath": "/web/app/themes/child"
    },
    {
      "format": "compressed",
      "extensionName": ".min.css",
      "savePath": "/web/app/themes/child"
    },
  ],
````

It would be better to follow a common rule for path formats in the options.
Personally, I like the second one (`savePath`) better, for me, it's more straightforward.

**Comment by glenn2223** [comment]
That's fine, consistency is everything 😄

I'll push this in v5 because, in my haste, it is a breaking change. My concern was around the user entering a file path rather than a folder. I've got a solution for that though

***Note***
```TS
try{ if (!fs.statSync(baseDirectory).isDirectory()) { /*err not folder*/ } } catch { /* err doesn't exist*/ }
```

**Comment by glenn2223** [comment]
Any input on #30 would be great 😄
